File: src/domain/editLearning.ts
import type {
  PostRevisionAuthorship,
  PostRevisionOrigin,
} from "./posts";

export type EditChangeKind = "addition" | "removal" | "replacement";
export type EditLearningDecisionKind = "accepted" | "rejected";

export interface EditChangeView {
  kind: EditChangeKind;
  removedText: string | null;
  addedText: string | null;
}

export interface EditObservationView {
  postId: string;
  parentRevisionId: string;
  childRevisionId: string;
  parentOrigin: PostRevisionOrigin;
  parentAuthorshipState: PostRevisionAuthorship;
  childOrigin: PostRevisionOrigin;
  childAuthorshipState: PostRevisionAuthorship;
  changes: EditChangeView[];
}

export interface EditPairProvenanceView {
  postId: string;
  parentRevisionId: string;
  childRevisionId: string;
}

export interface RecurringEditPreferenceProposalView {
  proposalKey: string;
  kind: EditChangeKind;
  statement: string;
  supportCount: number;
  distinctPostCount: number;
  supportingPairs: EditPairProvenanceView[];
}

export interface DecideEditLearningProposalRequest {
  proposalKey: string;
  supportingPairs: EditPairProvenanceView[];
  decision: EditLearningDecisionKind;
}

export interface EditLearningDecisionView {
  decisionId: string;
  proposalFingerprint: string;
  decision: EditLearningDecisionKind;
  supportingPairs: EditPairProvenanceView[];
  resultingArtifactType: string | null;
  resultingArtifactId: string | null;
  decidedAt: string;
}

export interface EditLearningAnalysisView {
  observations: EditObservationView[];
  proposals: RecurringEditPreferenceProposalView[];
}

File: src-tauri/src/domain/edit_learning.rs
use serde::{Deserialize, Serialize};

use crate::domain::posts::{PostRevisionAuthorship, PostRevisionOrigin};

#[derive(Debug, Clone, Copy, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EditChangeKind {
    Addition,
    Removal,
    Replacement,
}

impl EditChangeKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Addition => "addition",
            Self::Removal => "removal",
            Self::Replacement => "replacement",
        }
    }
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EditChangeView {
    pub kind: EditChangeKind,
    pub removed_text: Option<String>,
    pub added_text: Option<String>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EditObservationView {
    pub post_id: String,
    pub parent_revision_id: String,
    pub child_revision_id: String,
    pub parent_origin: PostRevisionOrigin,
    pub parent_authorship_state: PostRevisionAuthorship,
    pub child_origin: PostRevisionOrigin,
    pub child_authorship_state: PostRevisionAuthorship,
    pub changes: Vec<EditChangeView>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EditPairProvenanceView {
    pub post_id: String,
    pub parent_revision_id: String,
    pub child_revision_id: String,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RecurringEditPreferenceProposalView {
    pub proposal_key: String,
    pub kind: EditChangeKind,
    pub statement: String,
    pub support_count: usize,
    pub distinct_post_count: usize,
    pub supporting_pairs: Vec<EditPairProvenanceView>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EditLearningDecisionKind {
    Accepted,
    Rejected,
}

impl EditLearningDecisionKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Accepted => "accepted",
            Self::Rejected => "rejected",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DecideEditLearningProposalRequest {
    pub proposal_key: String,
    pub supporting_pairs: Vec<EditPairProvenanceView>,
    pub decision: EditLearningDecisionKind,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EditLearningDecisionView {
    pub decision_id: String,
    pub proposal_fingerprint: String,
    pub decision: EditLearningDecisionKind,
    pub supporting_pairs: Vec<EditPairProvenanceView>,
    pub resulting_artifact_type: Option<String>,
    pub resulting_artifact_id: Option<String>,
    pub decided_at: String,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EditLearningAnalysisView {
    pub observations: Vec<EditObservationView>,
    pub proposals: Vec<RecurringEditPreferenceProposalView>,
}
